### PR TITLE
[DNM] FHIR Gateway Enhancements

### PR DESCRIPTION
**Pull request by ndegwamartin**
**IMPORTANT: Where possible all PRs must be linked to a Github issue**

**In this PR**
- Fix REL sync strategy Caching bug
- Refactor to support returning all paginated results
- Refactor REL upstream fetch to single request
- Pagination count bumped to 200
- Release version 2.1.2

**Engineer Checklist**

- [ ] I have written **Unit tests** for any new feature(s) and edge cases for
      bug fixes
- [ ] I have added documentation for any new feature(s) and configuration
      option(s) on the `README.md`
- [ ] I have run `mvn spotless:check` to check my code follows the project's
      style guide
- [ ] I have run `mvn clean test jacoco:report` to confirm the coverage report
      was generated at `plugins/target/site/jacoco/index.html`
- [ ] I ran `mvn clean package` right before creating this pull request.

### CODE REVIEW

File: exec/pom.xml [hunk 1]
```diff
@@ -4,7 +4,7 @@
   <parent>
     <groupId>org.smartregister</groupId>
     <artifactId>opensrp-gateway-plugin</artifactId>
-    <version>2.1.1</version>
+    <version>2.1.2.0</version>
```
Comment (dubdabasoduba) [inline exec/pom.xml:7]: Update to the correct version, we already have `2.1.2.3` deployed.

File: exec/pom.xml [hunk 2]
```diff
@@ -70,7 +70,7 @@
     <dependency>
       <groupId>org.smartregister</groupId>
       <artifactId>plugins</artifactId>
-      <version>2.1.1</version>
+      <version>2.1.2.0</version>
```
Comment (dubdabasoduba) [inline exec/pom.xml:73]: Same as above

File: plugins/pom.xml [hunk 3]
```diff
@@ -4,7 +4,7 @@
   <parent>
     <groupId>org.smartregister</groupId>
     <artifactId>opensrp-gateway-plugin</artifactId>
-    <version>2.1.1</version>
+    <version>2.1.2.0</version>
```
Comment (dubdabasoduba) [inline plugins/pom.xml:7]: Same as above

File: pom.xml [hunk 4]
```diff
@@ -11,7 +11,7 @@
 
   <groupId>org.smartregister</groupId>
   <artifactId>opensrp-gateway-plugin</artifactId>
-  <version>2.1.1</version>
+  <version>2.1.2.0</version>
```
Comment (dubdabasoduba) [inline pom.xml:14]: Same as above

**Review by dubdabasoduba — CHANGES_REQUESTED** [review]

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/PermissionAccessChecker.java [hunk 5]
```diff
@@ -129,6 +131,35 @@ private void initSyncAccessDecision(RequestDetailsReader requestDetailsReader) {
...
+        switch (syncStrategy) {
+            case Constants.SyncStrategy.RELATED_ENTITY_LOCATION:
+                try {
+
+                    String[] syncLocations =
+                            parameters.getOrDefault(
+                                    Constants.SYNC_LOCATIONS_SEARCH_PARAM, new String[] {});
+
+                    if (syncLocations.length == 0)
+                        throw new IllegalStateException("No _syncLocation query param specified");
```
[the remark this answers is not in the document]
Comment (ndegwamartin) [reply plugins/src/main/java/org/smartregister/fhir/gateway/plugins/PermissionAccessChecker.java]: @dubdabasoduba how should the Gateway respond for requests that are using REL but do not have _syncLocation specified?

[the remark this answers is not in the document]
Comment (dubdabasoduba) [reply plugins/src/main/java/org/smartregister/fhir/gateway/plugins/PermissionAccessChecker.java]: It should return an error saying that the sync locations are not set.  However, We should ensure the requests made from the composition resources don't throw that error. Those requests are not affected by the sync strategies.

[the remark this answers is not in the document]
Comment (ndegwamartin) [reply plugins/src/main/java/org/smartregister/fhir/gateway/plugins/PermissionAccessChecker.java]: I think those coming from composition should be ignored on the Allowed Queries or the Ignore Sync filter configuration file.
